plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -311,62 +395,101 @@ func (c *CloudWatch) fetchNamespaceMetrics() ([]*cloudwatch.Metric, error) {
 /*
  * Gather given Metric and emit any error
  */
-func (c *CloudWatch) gatherMetric(
-	acc telegraf.Accumulator,
-	metric *cloudwatch.Metric,
-) error {
-	params := c.getStatisticsInput(metric)
-	resp, err := c.client.GetMetricStatistics(params)
+func (c *CloudWatch) gatherMetrics(
+	params *cloudwatch.GetMetricDataInput,
+) ([]*cloudwatch.MetricDataResult, error) {
+	resp, err := c.client.GetMetricData(params)
 	if err != nil {
-		return err
+		return nil, errors.New("Failed to get metric data - " + err.Error())
+	}
+
+	// todo: determine if we need to handle pagination.
+	if resp.NextToken != nil {
+		log.Println("[inputs.cloudwatch] W! UNHANDLED PAGINATED RESULTS!!", *resp.NextToken)
+	}
+	return resp.MetricDataResults, nil
+}
+
+func aggregateFiltered(filteredMetrics []filteredMetric) []*cloudwatch.Metric {
+	metrics := []*cloudwatch.Metric{}
+	for _, filtered := range filteredMetrics {
+		metrics = append(metrics, filtered.metrics...)
 	}
+	return metrics
+}
 
-	for _, point := range resp.Datapoints {
+func (c *CloudWatch) aggregateMetrics(
+	acc telegraf.Accumulator,
+	metrics []*cloudwatch.Metric,
+	metricDataResults []*cloudwatch.MetricDataResult,
+) error {
+	for _, metric := range metrics {
+		timestamp := time.Now()
+		fields := map[string]interface{}{}
 		tags := map[string]string{
 			"region": c.Region,
-			"unit":   snakeCase(*point.Unit),
-		}
-
-		for _, d := range metric.Dimensions {
-			tags[snakeCase(*d.Name)] = *d.Value
 		}
 
-		// record field for each statistic
-		fields := map[string]interface{}{}
+		// todo: group these once beforehand if possible to avoid looping through all results metrics number of times
+		results := getResults(*metric, metricDataResults)
+		for _, result := range results {
+			for _, dimension := range metric.Dimensions {
+				tags[snakeCase(*dimension.Name)] = *dimension.Value
+			}
+			if len(result.Timestamps) == 0 || result.Timestamps[0] == nil {
+				continue
+			}
+			// todo: determine if we need to handle multiple values.
```

Comment:
maybe? https://github.com/influxdata/telegraf/issues/5420#issuecomment-475783462 :)

Reply:
I'm not certain we'd gain any benefit from parsing these. I think I'm going to stop logging that and just always take the first value. If we looped through, it would take the last value as the field name will always be the same. Since the timestamp is re-updated on each result, it would have the same timestamp as well.

Reply:
I believe this is something we need to support or we would break  https://github.com/influxdata/telegraf/issues/4643.